Issue #13934: Version updated from 3.2 to 3.4

Rohanraj123 opened the pull request:
related #13934

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ReturnCountCheck.java:
```diff
@@ -164,7 +164,7 @@ public void setFormat(Pattern pattern) {
      *
      * @param max maximum allowed number of return statements.
-     * @since 3.2
+     * @since 3.4
```
See previous comment. You can see max setter in it.

rnveach commented:
From issue: https://github.com/checkstyle/checkstyle/pull/13932/files#r1368105661

Property was `format` originally. It was changed from 3.4 to 3.2 originally.

https://github.com/checkstyle/checkstyle/blob/release3_4/src/checkstyle/com/puppycrawl/tools/checkstyle/checks/AbstractFormatCheck.java

3.2 check above didn't specify `format` or extend `AbstractFormatCheck`.

It did extend `AbstractFormatCheck` in 3.4 .
https://github.com/checkstyle/checkstyle/blob/release3_4/src/checkstyle/com/puppycrawl/tools/checkstyle/checks/coding/ReturnCountCheck.java

But it did in 3.4 .
https://github.com/checkstyle/checkstyle/blob/release3_4/src/checkstyle/com/puppycrawl/tools/checkstyle/checks/AbstractFormatCheck.java

Rohanraj123 commented:
@romani Sir can you please review the changes and please provide me some enhancements if required.

rnveach commented:
@Rohanraj123 You have not resolved all my review items yet.

romani commented:
@Rohanraj123 , please redirect all questions to @rnveach , he is first reviewer for this PR.

Rohanraj123 commented:
> @Rohanraj123 You have not resolved all my review items yet.

@rnveach Sir, can you please one more time define the doables so that i can keep them in mind while making the changes. Would be helpful for my learning as well.

rnveach commented:
@Rohanraj123 Where is the confusion with the information you were given at https://github.com/checkstyle/checkstyle/issues/13934#issuecomment-1777352322 and my review items at https://github.com/checkstyle/checkstyle/pull/13945#discussion_r1370267758 , https://github.com/checkstyle/checkstyle/pull/13945#discussion_r1370270679 , and https://github.com/checkstyle/checkstyle/pull/13945#issuecomment-1777336667 .

rnveach commented:
@Rohanraj123 
The line at https://github.com/checkstyle/checkstyle/blob/0322ba1be8489d1df373013efc9078de4d209eb4/src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java#L702 needs to be removed.

With it gone, it will re-generate the XML files with the correct versions.

Rohanraj123 commented:
> @Rohanraj123 The line at
> 
> https://github.com/checkstyle/checkstyle/blob/0322ba1be8489d1df373013efc9078de4d209eb4/src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java#L702
> 
> needs to be removed.
> With it gone, it will re-generate the XML files with the correct versions. I recommend to start with a clean master first.

Sure sir Let me do the changes

Rohanraj123 commented:
@rnveach Sir, removing the line of code you were reffering giving error : ```
[ERROR] com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest -- Time elapsed: 2.495 s <<< ERROR!
java.lang.NullPointerException: Cannot invoke "com.puppycrawl.tools.checkstyle.api.DetailNode.getChildren()" because "node" is null
```

rnveach commented:
Update the PR. I would need to see what is being done and the full stack trace.

Rohanraj123 commented:
@rnveach Sir, I have added the catch check in XdocGenerator file. And squash them in one commit. When i tried to run mvn clean verify , It says that Exception arrives in src\xdocs\checks\whitespace\filetabcharacter.xml.template .
Sir it might take some time for me to work on this issue because my university exams are going on. So Please try to provide me some time.

romani reviewed (approved):
Ok to merge

rnveach commented:
@Rohanraj123 No worries. You have some time. I am looking into the issue a bit more to understand things.

@romani CI is failing, so is too soon to say it is ready for merging.

rnveach commented:
I was digging through the error and I think I understand the logic of why the code we removed is there. `filetabcharacter` is an `FileSetCheck`. The error for the since version is because it is looking for the javadoc of `fileExtensions` and not finding it since it is in the parent class. This is a property for all `FileSetCheck`, so I assume the reasoning for the removed code was that this class has been here long enough that all modules that use it, should have had it from the start. However, this logic starts to fall apart if we add a new property in the future to any abstract checks that are already in existence.

We can actually see this for 2 of the overrides we did for `SINCE_VERSION_FOR_INHERITED_PROPERTY`:
https://github.com/checkstyle/checkstyle/blob/1282fe6a9e07a7b2ec76e0a86144e55899294c35/src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java#L165-L166

And I think that's where the solution lies, we need to add these properties to this global override field. So I would expect us to add a few `<ModuleName>.fileExtensions` to the list with each module's specific version that it defines. Most likely all may be the same as Module's own version number.

Ideally, we could remove the need for this list if we did version comparing between the version the module was introduced versus the version the inherited property was introduced and taking the newest of the 2. We are tracking inherited properties directly through `SUPER_CLASS_PROPERTIES_JAVADOCS`. I would expect this to be done in another PR. @romani If you agree we should create the issue for this.

romani commented:
@Rohanraj123, if you think that this PR becomes a bit complicated to finish.
It is completely fine to pause it or abandon it and switch to easier issues to get experience a bit more before handling complicated.
We want you to be productive and not be burned by heavy problems.

Rohanraj123 commented:
> I was digging through the error and I think I understand the logic of why the code we removed is there. `filetabcharacter` is an `FileSetCheck`. The error for the since version is because it is looking for the javadoc of `fileExtensions` and not finding it since it is in the parent class and has no version number defined there. This is a property for all `FileSetCheck`, so I assume the reasoning for the removed code was that this class has been here long enough that all modules that use it, should have had it from the start. However, this logic starts to fall apart if we add a new property in the future to any abstract checks that are already in existence.
> 
> We can actually see this for 2 of the overrides we did for `SINCE_VERSION_FOR_INHERITED_PROPERTY` for 2 same inherited properties:
> 
> https://github.com/checkstyle/checkstyle/blob/1282fe6a9e07a7b2ec76e0a86144e55899294c35/src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java#L165-L166
> 
> And I think that's where the solution lies, we need to add these properties to this global override field. So I would expect us to add a few `<ModuleName>.fileExtensions` to the list with each module's specific version that it defines. Most likely all may be the same as Module's own version number.
> 
> Ideally, we could remove the need for this list if we did version comparing between the version the module was introduced versus the version the inherited property was introduced and taking the newest of the 2. We are tracking inherited properties directly through `SUPER_CLASS_PROPERTIES_JAVADOCS`. I would expect this to be done in another PR. @romani If you agree we should create the issue for this.
> 
> Edit: We would still need the override to fall back on if we changed extension of the module after it was introduced, like changing AbstractCheck to AbstractJavadocCheck. We have a few modules we are looking to do that to still. Only way to avoid the override is if we put a `since` version on the class hierarchy.

Sure @rnveach Sir , letme add <ModuleName>.fileextensions first. Then we will discuss furthur

romani commented:
ci failure:
```
[INFO] Running com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.296 s <<< FAILURE! -- in com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest
[ERROR] com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest -- Time elapsed: 2.296 s <<< ERROR!
java.lang.IllegalStateException: Exception was thrown while processing src/xdocs/checks/whitespace/typecastparenpad.xml.template
  at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
 ....
  at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
Caused by: java.lang.NullPointerException
  at com.puppycrawl.tools.checkstyle.utils.JavadocUtil.getFirstChild(JavadocUtil.java:220)
  at com.puppycrawl.tools.checkstyle.utils.JavadocUtil.findFirstToken(JavadocUtil.java:200)
  at com.puppycrawl.tools.checkstyle.site.SiteUtil.getSinceVersionFromJavadoc(SiteUtil.java:727)
  at com.puppycrawl.tools.checkstyle.site.SiteUtil.getSinceVersion(SiteUtil.java:707)
  at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writePropertySinceVersionCell(PropertiesMacro.java:492)
  at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writePropertyRow(PropertiesMacro.java:233)
  at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writeTablePropertiesRows(PropertiesMacro.java:187)
  at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writePropertiesTable(PropertiesMacro.java:130)
  at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.execute(PropertiesMacro.java:102)
  at org.apache.maven.doxia.parser.AbstractParser.executeMacro(AbstractParser.java:145)
  at com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser.processMacroEnd(XdocsTemplateParser.java:187)
  at com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser.handleEndTag(XdocsTemplateParser.java:116)
  at org.apache.maven.doxia.parser.AbstractXmlParser.parseXml(AbstractXmlParser.java:238)
  at org.apache.maven.doxia.parser.AbstractXmlParser.parse(AbstractXmlParser.java:156)
  at org.apache.maven.doxia.parser.XhtmlBaseParser.parse(XhtmlBaseParser.java:92)
  at org.apache.maven.doxia.module.xdoc.XdocParser.parse(XdocParser.java:113)
  at com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser.parse(XdocsTemplateParser.java:76)
  at org.apache.maven.doxia.parser.AbstractParser.parse(AbstractParser.java:207)
  at com.puppycrawl.tools.checkstyle.internal.utils.XdocGenerator.generateXdocContent(XdocGenerator.java:68)
  ... 62 more

```

Rohanraj123 commented:
> ci failure:
> 
> ```
> [INFO] Running com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest
> [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.296 s <<< FAILURE! -- in com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest
> [ERROR] com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest -- Time elapsed: 2.296 s <<< ERROR!
> java.lang.IllegalStateException: Exception was thrown while processing src/xdocs/checks/whitespace/typecastparenpad.xml.template
>   at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
>  ....
>   at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
> Caused by: java.lang.NullPointerException
>   at com.puppycrawl.tools.checkstyle.utils.JavadocUtil.getFirstChild(JavadocUtil.java:220)
>   at com.puppycrawl.tools.checkstyle.utils.JavadocUtil.findFirstToken(JavadocUtil.java:200)
>   at com.puppycrawl.tools.checkstyle.site.SiteUtil.getSinceVersionFromJavadoc(SiteUtil.java:727)
>   at com.puppycrawl.tools.checkstyle.site.SiteUtil.getSinceVersion(SiteUtil.java:707)
>   at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writePropertySinceVersionCell(PropertiesMacro.java:492)
>   at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writePropertyRow(PropertiesMacro.java:233)
>   at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writeTablePropertiesRows(PropertiesMacro.java:187)
>   at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writePropertiesTable(PropertiesMacro.java:130)
>   at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.execute(PropertiesMacro.java:102)
>   at org.apache.maven.doxia.parser.AbstractParser.executeMacro(AbstractParser.java:145)
>   at com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser.processMacroEnd(XdocsTemplateParser.java:187)
>   at com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser.handleEndTag(XdocsTemplateParser.java:116)
>   at org.apache.maven.doxia.parser.AbstractXmlParser.parseXml(AbstractXmlParser.java:238)
>   at org.apache.maven.doxia.parser.AbstractXmlParser.parse(AbstractXmlParser.java:156)
>   at org.apache.maven.doxia.parser.XhtmlBaseParser.parse(XhtmlBaseParser.java:92)
>   at org.apache.maven.doxia.module.xdoc.XdocParser.parse(XdocParser.java:113)
>   at com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser.parse(XdocsTemplateParser.java:76)
>   at org.apache.maven.doxia.parser.AbstractParser.parse(AbstractParser.java:207)
>   at com.puppycrawl.tools.checkstyle.internal.utils.XdocGenerator.generateXdocContent(XdocGenerator.java:68)
>   ... 62 more
> ```

I fixed this CI failure @romani Sir.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java:
```diff
@@ -699,9 +702,7 @@ public static String getSinceVersion(String moduleName, DetailNode moduleJavadoc
-        else if (SUPER_CLASS_PROPERTIES_JAVADOCS.containsKey(propertyName)
-                || TOKENS.equals(propertyName)
-                || JAVADOC_TOKENS.equals(propertyName)) {
+        else if (TOKENS.equals(propertyName) || JAVADOC_TOKENS.equals(propertyName)) {
```
````suggestion
        else if (TOKENS.equals(propertyName)
                || JAVADOC_TOKENS.equals(propertyName)) {
````

Rohanraj123 replied:
Sure sir

Rohanraj123 replied:
@rnveach Even After making these changes it is not updating sinceVersion.

romani replied:
@Rohanraj123 , restore formatting only
```
else if (TOKENS.equals(propertyName)
                || JAVADOC_TOKENS.equals(propertyName)) {
```

romani replied:
code looks good.

rnveach replied:
> Even After making these changes it is not updating sinceVersion.

What do you mean?
It was updated at https://github.com/checkstyle/checkstyle/pull/13945/files#diff-eaff5fa3cb101145757a2bd9ebe2e805e8fd8057ae27f9e21ef91900cb7938cc

Rohanraj123 replied:
Okay got it thanks !

Rohanraj123 replied:
> @Rohanraj123 , restore formatting only
> 
> ```
> else if (TOKENS.equals(propertyName)
>                 || JAVADOC_TOKENS.equals(propertyName)) {
> ```

did it ! @romani Sir

rnveach reviewed (changes requested):
Should be final change
